### DISCUSSION TITLE
Getting Multiple Fields Using array_agg Function

### DISCUSSION THREAD

**Discussion by kiwavi** [sqlalchemy/sqlalchemy#11156] [Usage Questions] [answered]
Hello,

I have the following query in sql:

` SELECT orders.payment_id AS payment_id, array_agg(orders) AS array_agg_1 FROM orders where paid=true GROUP BY orders.payment_id;`

It rightfully returns, as an array, all the rows that have the same payment_id. 

Using sqlAlchemy, the following query works:

`db.query(models.Orders.payment_id,func.array_agg(models.Orders.payment_id)).group_by(models.Orders.payment_id).all()`

I have to only use one field in the array_agg expression parameter for it to work. Issue is that it returns only that one field while I need all field. 

If i use the query above with the payment_id in the array_agg ommited, i get an error saying sqlalchemy.exc.ArgumentError: SQL expression element expected, got <class 'Utils.models.Orders'>, which I understand. What kind of expression (that does not involve concatenation) would return all fields?

Thanks in advance

**Comment by CaselIT** [comment]
Hi,

So what you are seeking it to return a list of instances of Orders in your array agg? Something like this
`[(1, [Orders(), Orders()]), (2, [Orders(), ...])]`?

**Reply by kiwavi** [reply]
Yeah something like that. With the orders fully populated, instead of only one field. Actually I'm achieving that with raw sql. With sqlalchemy is where the challenge comes.

**Reply by CaselIT** [reply]
If you don't need orm instaces in the array-agg column that's doable: using `func.array_agg(Orders.__table__.table_valued())` should work

**Reply by kiwavi** [reply]
I tried the command. While it brings all fields now, each character is enclosed in quotations. Example:
```

"{",
        "\"",
        "(",
        "1",
        ",",
        ",",
        "d",
        "m",
        "u",
        "n",
```

**Reply by CaselIT** [reply, marked as answer]
Right that's not supported by most dialects, see also https://github.com/sqlalchemy/sqlalchemy/discussions/9951. Forgot about that

Using `json_agg` works though

**Reply by kiwavi** [reply]
Yeah this works well. Much appreciated.

**Comment by zzzeek** [comment]
can you show us an example of the ORM model you are working with ?   what is the array_agg achieving here?  it's not clear what you are actually intending and the array_agg seems unnecessary

**Reply by kiwavi** [reply]
Yeah. So I have an orders table. Each order has a payment_id which is a foreign key to a Payments table. Some orders have the same payment_id. I want to aggregate these orders and associate them with this payment_id. I hope that's clear enough. 

```
class Orders(Base):
    # represents each checkout session
    __tablename__ = 'orders'

    id = Column(Integer, primary_key=True,index=True)
    customer_name = Column(String)
    customer_email = Column(String)
    location = Column(String,nullable=False)
    creation_date = Column(DateTime, default=datetime.now())
    deletion_date = Column(DateTime)
    total_price = Column(Integer)
    product_details = Column(String)
    delivery_id = Column(Integer,ForeignKey("deliveries.id"),nullable=False)
    bag_id = Column(Integer,ForeignKey("bags.id"))
    paid = Column(Boolean,default=False)
    quantity = Column(Integer)
    order_status = Column(String,default="pending")
    customer_number = Column(String,nullable=False)
    payment_id = Column(Integer,ForeignKey("payments.id"))
```

```

class Payments(Base):
    __tablename__ = 'payments'
    
    id = Column(Integer,primary_key=True,index=True)
    amount = Column(Integer)
    creation_date = Column(DateTime,default=datetime.now())
    deletion_date = Column(DateTime)
    CheckoutRequestID = Column(String)
    MerchantRequestID = Column(String)
    MpesaReceiptNumber = Column(String)
    ResponseCode = Column(Integer)
    ResultDesc = Column(String)
    log = Column(String)
    ResultCode = Column(Integer)
    PhoneNumber = Column(String)
    TransactionDate = Column(String)
```

**Reply by CaselIT** [reply]
I'm not sure you can do that using a query, but relationship does exactly what you are looking for in an object oriented way.
See https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html and eager loads https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html
